File: Heap/1792. Maximum Average Pass Ratio/solution.py
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        maxHeap = [] # gain, p, t
        for p, t in classes:
            nxtGain = (p+1)/(t+1) - p/t
            heapq.heappush(maxHeap, [-nxtGain, p, t])

        while extraStudents:
            _, p, t = heapq.heappop(maxHeap)
            p, t = p+1, t+1
            extraStudents -= 1

            nxtGain = (p+1)/(t+1) - p/t
            heapq.heappush(maxHeap, [-nxtGain, p, t])
        
        return sum(p/t for _, p, t in maxHeap)/len(maxHeap)

File: Heap/1792. Maximum Average Pass Ratio/solution.cpp
using pdi = pair<double,int>;

class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        int n = classes.size();
        priority_queue<pdi> pq; // max heap
        for (int i=0; i<n; i++) {
            double pass = classes[i][0], total = classes[i][1];
            double next_gain = ((pass+1) / (total+1)) - pass/total;
            pq.push({next_gain, i});
        }

        for (int i=0; i<extraStudents; i++) {
            int idx = pq.top().second;
            pq.pop();
            
            classes[idx][0]++;
            classes[idx][1]++;
            double pass = classes[idx][0], total = classes[idx][1];
            double next_gain = ((pass+1) / (total+1)) - pass/total;
            
            pq.push({next_gain, idx});
        }

        double res = 0;
        while (!pq.empty()) {
            int i = pq.top().second;
            pq.pop();
            res += (double) classes[i][0]/classes[i][1];
        }
        return res / n;
    }
};
